File: packages/protocol/src/metabolism.ts
import { digestJson, isSha256Digest, type JsonValue } from "./canonical.js";
import type { DsseEnvelope } from "./types.js";

export const METABOLIC_KINDS = ["Mass", "Refraction", "Polarity", "Fission", "Inertia"] as const;
export type MetabolicKind = typeof METABOLIC_KINDS[number];
export const METABOLIC_RESOURCE_KEYS = ["maxMindInvocations", "maxInputTokens", "maxOutputTokens", "maxWallMillis", "maxGeneratedBytes", "maxForgeCpuMillis", "maxForgeWallMillis", "maxMemorySeconds", "maxWritableBytes", "maxWritableInodes", "maxArtifactBytes", "maxTotalAssayCost"] as const;
export type MetabolicResourceVector = Readonly<Record<typeof METABOLIC_RESOURCE_KEYS[number], number>>;
export const METABOLIC_WORK_KEYS = ["general", "unfamiliarFamilies", "counterbelief", "isolatedLanes", "scentContinuation"] as const;
export const METABOLIC_WORK_FOR_KIND = Object.freeze({ Mass: "general", Refraction: "unfamiliarFamilies", Polarity: "counterbelief", Fission: "isolatedLanes", Inertia: "scentContinuation" } as const);
export type MetabolicWork = Readonly<Record<typeof METABOLIC_WORK_KEYS[number], number>>;
export const METABOLISM_DSSE_PAYLOAD_TYPE = "application/vnd.jevyr.metabolism-receipt+json";

export interface MetabolicInterval { readonly lower: number; readonly upper: number; readonly confidence: 0.95 }
export interface MetabolicCalibrationSummary {
  readonly digest: string;
  readonly scope: string;
  readonly pairedSeeds: number;
  readonly doses: readonly number[];
  readonly monotonic: true;
  readonly identicalEvidenceJudgment: true;
  readonly recallDifference: MetabolicInterval;
  readonly reproducibilityDifference: MetabolicInterval;
  readonly nonInferiorityMargin: number;
  readonly maximumObservedUnitResources: MetabolicResourceVector;
}
export interface MetabolicAllowance {
  readonly protocol: "jevyr.metabolic-allowance/1";
  readonly baselineSearchDigest: string;
  readonly maximumAddedResources: MetabolicResourceVector;
  readonly maximumQuantityPerBall: number;
  readonly maximumReceipts: number;
  readonly unitCostCeilings: Readonly<Record<MetabolicKind, MetabolicResourceVector>>;
  readonly calibrationSetDigest: string;
  readonly calibrations: readonly { readonly kind: MetabolicKind; readonly digest: string; readonly summaryDigest: string; readonly maximumDose: number }[];
  readonly signer: { readonly keyId: string; readonly publicKeyPem: string };
  readonly digest: string;
}
export interface MetabolicOffer {
  readonly ballId: string;
  readonly kind: MetabolicKind;
  readonly minQuantity: 1;
  readonly maxQuantity: number;
  readonly promisedEffect: string;
  readonly unitCostCeiling: MetabolicResourceVector;
  readonly unitWork: MetabolicWork;
  readonly calibration: MetabolicCalibrationSummary;
}
export interface MetabolicOffers {
  readonly protocol: "jevyr.metabolic-offers/1";
  readonly caseId: string;
  readonly runDigest: string;
  readonly allowanceDigest: string;
  readonly baselineSearchDigest: string;
  readonly admission: "open" | "closed";
  readonly cumulativeGrant: MetabolicResourceVector;
  readonly cumulativeQuantities: Readonly<Record<MetabolicKind, number>>;
  readonly offers: readonly MetabolicOffer[];
  readonly receipts: readonly MetabolicRedemption[];
}
export interface MetabolicReceipt {
  readonly protocol: "jevyr.metabolism-receipt/1";
  readonly caseId: string;
  readonly caseDigest: string;
  readonly runDigest: string;
  readonly policyDigest: string;
  readonly searchDigest: string;
  readonly allowanceDigest: string;
  readonly calibrationDigest: string;
  readonly ballId: string;
  readonly kind: MetabolicKind;
  readonly quantity: number;
  readonly sequence: number;
  readonly previousReceiptDigest: string | null;
  readonly issuedAt: string;
  readonly grant: MetabolicResourceVector;
  readonly cumulativeGrant: MetabolicResourceVector;
  readonly cumulativeQuantities: Readonly<Record<MetabolicKind, number>>;
  readonly work: MetabolicWork;
  readonly baselineUnchanged: true;
  readonly verdictAuthority: "none";
  readonly digest: string;
}
export interface MetabolicRedemption {
  readonly protocol: "jevyr.metabolism-redemption/1";
  readonly receipt: MetabolicReceipt;
  readonly envelope: DsseEnvelope;
}

function object(value: unknown, keys: readonly string[], label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value) || ![Object.prototype, null].includes(Object.getPrototypeOf(value))) throw new TypeError(`${label} must be an ordinary object`);
  const descriptors = Object.getOwnPropertyDescriptors(value);
  if (Reflect.ownKeys(value).length !== keys.length || keys.some((key) => !Object.hasOwn(descriptors, key)) || Object.values(descriptors).some((entry) => !entry.enumerable || !Object.hasOwn(entry, "value"))) throw new TypeError(`${label} has an unknown, missing, or accessor field`);
  return value as Record<string, unknown>;
}
function integer(value: unknown, min: number, max: number, label: string): asserts value is number {
  if (!Number.isSafeInteger(value) || Number(value) < min || Number(value) > max) throw new TypeError(`${label} must be an integer from ${min} through ${max}`);
}
function text(value: unknown, maximum: number, label: string): asserts value is string {
  if (typeof value !== "string" || value.length < 1 || value.length > maximum || /[\u0000-\u001f\u007f]/u.test(value)) throw new TypeError(`${label} must be bounded text`);
}
export function assertMetabolicResourceVector(value: unknown): asserts value is MetabolicResourceVector {
  const vector = object(value, METABOLIC_RESOURCE_KEYS, "Metabolic resources");
  for (const key of METABOLIC_RESOURCE_KEYS) integer(vector[key], 0, Number.MAX_SAFE_INTEGER, key);
}
export function assertMetabolicWork(value: unknown): asserts value is MetabolicWork {
  const work = object(value, METABOLIC_WORK_KEYS, "Metabolic work");
  for (const key of METABOLIC_WORK_KEYS) integer(work[key], 0, 1_000_000, key);
}
export function assertMetabolicAllowance(value: unknown): asserts value is MetabolicAllowance {
  const allowance = object(value, ["protocol", "baselineSearchDigest", "maximumAddedResources", "maximumQuantityPerBall", "maximumReceipts", "unitCostCeilings", "calibrationSetDigest", "calibrations", "signer", "digest"], "Metabolic allowance");
  if (allowance.protocol !== "jevyr.metabolic-allowance/1") throw new TypeError("Unknown metabolic allowance protocol");
  for (const key of ["baselineSearchDigest", "calibrationSetDigest", "digest"]) if (!isSha256Digest(allowance[key])) throw new TypeError(`Invalid metabolic ${key}`);
  integer(allowance.maximumQuantityPerBall, 1, 64, "maximumQuantityPerBall");
  integer(allowance.maximumReceipts, 1, 4096, "maximumReceipts");
  assertMetabolicResourceVector(allowance.maximumAddedResources);
  if (!Array.isArray(allowance.calibrations) || allowance.calibrations.length > METABOLIC_KINDS.length) throw new TypeError("Metabolic calibration set is unbounded");
  const kinds = new Set<string>();
  for (const calibration of allowance.calibrations) {
    const item = object(calibration, ["kind", "digest", "summaryDigest", "maximumDose"], "Metabolic calibration binding");
    if (!METABOLIC_KINDS.includes(item.kind as MetabolicKind) || kinds.has(String(item.kind)) || !isSha256Digest(item.digest) || !isSha256Digest(item.summaryDigest)) throw new TypeError("Metabolic calibration kind is repeated or invalid");
    integer(item.maximumDose, 2, 64, "maximumDose"); kinds.add(String(item.kind));
  }
  const reports = allowance.calibrations.map((item: { kind: string; digest: string }) => ({ kind: item.kind, digest: item.digest })).sort((a: {kind: string}, b: {kind: string}) => a.kind.localeCompare(b.kind));
  if (digestJson({ protocol: "jevyr.metabolic-calibration-set/1", reports }) !== allowance.calibrationSetDigest) throw new TypeError("Metabolic calibration set digest mismatch");
  const doses = object(allowance.unitCostCeilings, METABOLIC_KINDS, "Metabolic dose ceilings");
  for (const kind of METABOLIC_KINDS) {
    assertMetabolicResourceVector(doses[kind]);
    for (const key of METABOLIC_RESOURCE_KEYS) if (doses[kind][key] > allowance.maximumAddedResources[key]) throw new TypeError("A dose exceeds the sealed additive allowance");
  }
  const signer = object(allowance.signer, ["keyId", "publicKeyPem"], "Metabolic signer");
  if (!isSha256Digest(signer.keyId) || typeof signer.publicKeyPem !== "string" || signer.publicKeyPem.length > 1024 || !signer.publicKeyPem.startsWith("-----BEGIN PUBLIC KEY-----\n")) throw new TypeError("Invalid metabolic public signer");
  const { digest, ...body } = allowance;
  if (digestJson(body as JsonValue) !== digest) throw new TypeError("Metabolic allowance digest mismatch");
}
export function assertMetabolicReceipt(value: unknown): asserts value is MetabolicReceipt {
  const receipt = object(value, ["protocol", "caseId", "caseDigest", "runDigest", "policyDigest", "searchDigest", "allowanceDigest", "calibrationDigest", "ballId", "kind", "quantity", "sequence", "previousReceiptDigest", "issuedAt", "grant", "cumulativeGrant", "cumulativeQuantities", "work", "baselineUnchanged", "verdictAuthority", "digest"], "Metabolic receipt");
  if (receipt.protocol !== "jevyr.metabolism-receipt/1" || receipt.baselineUnchanged !== true || receipt.verdictAuthority !== "none" || !METABOLIC_KINDS.includes(receipt.kind as MetabolicKind)) throw new TypeError("Metabolic receipt changes semantic authority or has an unknown kind");
  text(receipt.caseId, 256, "caseId");
  if (typeof receipt.ballId !== "string" || !/^ball_[A-Za-z0-9_-]{43}$/u.test(receipt.ballId)) throw new TypeError("ballId must be a server-issued opaque identity");
  for (const key of ["caseDigest", "runDigest", "policyDigest", "searchDigest", "allowanceDigest", "calibrationDigest", "digest"]) if (!isSha256Digest(receipt[key])) throw new TypeError(`Invalid metabolic ${key}`);
  integer(receipt.quantity, 1, 64, "quantity"); integer(receipt.sequence, 1, 4096, "sequence");
  if (receipt.sequence === 1 ? receipt.previousReceiptDigest !== null : !isSha256Digest(receipt.previousReceiptDigest)) throw new TypeError("Metabolic receipt has an invalid prior link");
  if (typeof receipt.issuedAt !== "string" || !Number.isFinite(Date.parse(receipt.issuedAt)) || new Date(receipt.issuedAt).toISOString() !== receipt.issuedAt) throw new TypeError("Metabolic issuedAt must be canonical UTC time");
  assertMetabolicResourceVector(receipt.grant); assertMetabolicResourceVector(receipt.cumulativeGrant); assertMetabolicWork(receipt.work);
  const quantities = object(receipt.cumulativeQuantities, METABOLIC_KINDS, "Metabolic cumulative doses");
  for (const kind of METABOLIC_KINDS) integer(quantities[kind], 0, 64, `cumulativeQuantities.${kind}`);
  for (const key of METABOLIC_WORK_KEYS) if (receipt.work[key] !== (key === METABOLIC_WORK_FOR_KIND[receipt.kind as MetabolicKind] ? receipt.quantity : 0)) throw new TypeError("Metabolic work differs from its exact offered kind and quantity");
  for (const key of METABOLIC_RESOURCE_KEYS) if (receipt.grant[key] > receipt.cumulativeGrant[key]) throw new TypeError("Metabolic cumulative grant is smaller than its addition");
  const { digest, ...body } = receipt;
  if (digestJson(body as JsonValue) !== digest) throw new TypeError("Metabolic receipt digest mismatch");
}

/** Independently check every addition against the presealed allowance and prior link. */
export function assertMetabolicGrant(receipt: MetabolicReceipt, allowance: MetabolicAllowance, previous?: MetabolicReceipt): void {
  assertMetabolicReceipt(receipt); assertMetabolicAllowance(allowance);
  if (receipt.allowanceDigest !== allowance.digest || receipt.searchDigest !== allowance.baselineSearchDigest || receipt.quantity > allowance.maximumQuantityPerBall || receipt.sequence > allowance.maximumReceipts) throw new TypeError("Metabolic receipt exceeds or substitutes its sealed allowance");
  if (receipt.sequence !== (previous?.sequence ?? 0) + 1 || receipt.previousReceiptDigest !== (previous?.digest ?? null)) throw new TypeError("Metabolic receipt sequence has a gap or invalid predecessor");
  if (previous && ["caseId", "caseDigest", "runDigest", "policyDigest", "searchDigest", "allowanceDigest"].some((key) => receipt[key as keyof MetabolicReceipt] !== previous[key as keyof MetabolicReceipt])) throw new TypeError("Metabolic receipts cross a sealed identity boundary");
  const calibration = allowance.calibrations.find((entry) => entry.kind === receipt.kind);
  if (!calibration || calibration.digest !== receipt.calibrationDigest) throw new TypeError("Metabolic receipt uses an uncalibrated kind or substituted evidence");
  for (const kind of METABOLIC_KINDS) {
    const expected = (previous?.cumulativeQuantities[kind] ?? 0) + (kind === receipt.kind ? receipt.quantity : 0);
    if (receipt.cumulativeQuantities[kind] !== expected || expected > (allowance.calibrations.find((entry) => entry.kind === kind)?.maximumDose ?? 0)) throw new TypeError("Metabolic cumulative dose exceeds its measured calibration");
  }
  for (const key of METABOLIC_RESOURCE_KEYS) {
    if (BigInt(receipt.grant[key]) !== BigInt(allowance.unitCostCeilings[receipt.kind][key]) * BigInt(receipt.quantity)
      || receipt.cumulativeGrant[key] !== (previous?.cumulativeGrant[key] ?? 0) + receipt.grant[key]
      || receipt.cumulativeGrant[key] > allowance.maximumAddedResources[key]) throw new TypeError(`Metabolic ${key} exceeds its exact additive grant`);
  }
}

File: sdk/python/jevyr/metabolism.py
"""Live additive work, independently signed under the exact presealed allowance."""
from __future__ import annotations
import hashlib
import re
from typing import Any, Mapping
from .juggler import _check, _exact

KINDS = ("Mass", "Refraction", "Polarity", "Fission", "Inertia")
RESOURCES = ("maxMindInvocations", "maxInputTokens", "maxOutputTokens", "maxWallMillis", "maxGeneratedBytes", "maxForgeCpuMillis", "maxForgeWallMillis", "maxMemorySeconds", "maxWritableBytes", "maxWritableInodes", "maxArtifactBytes", "maxTotalAssayCost")
WORK = ("general", "unfamiliarFamilies", "counterbelief", "isolatedLanes", "scentContinuation")
PAYLOAD_TYPE = "application/vnd.jevyr.metabolism-receipt+json"

def _hash(value: Any) -> str:
    from .client import _canonical
    return "sha256:" + hashlib.sha256(_canonical(value)).hexdigest()

def _digest(value: Any) -> bool:
    return isinstance(value, str) and re.fullmatch(r"sha256:[a-f0-9]{64}", value) is not None

def _integer(value: Any, low: int = 0, high: int = 9007199254740991) -> bool:
    return type(value) is int and low <= value <= high

def _vector(value: Any, fields: tuple[str, ...] = RESOURCES) -> dict[str, Any]:
    result = _exact(value, set(fields), "Metabolic vector")
    _check(all(_integer(result[key]) for key in fields), "Metabolic vector contains an invalid bounded integer")
    return result

def assert_ball_id(value: Any) -> None:
    if not isinstance(value, str) or not re.fullmatch(r"ball_[A-Za-z0-9_-]{42}[AEIMQUYcgkosw048]", value):
        raise ValueError("Metabolism requires a canonical offered ballId")

def assert_quantity(value: Any) -> None:
    if not _integer(value, 1, 64):
        raise ValueError("Metabolism quantity must be an integer from 1 through 64")

def allowance_from_policy(policy: Any, seal: Mapping[str, Any]) -> dict[str, Any]:
    _check(policy.get("caseId") == seal["caseId"] and policy.get("policyDigest") == seal["policyDigest"], "Metabolic policy crosses its authenticated Seal")
    allowance = _exact(policy.get("artifact", {}).get("descriptor", {}).get("policy", {}).get("metabolicAllowance"), {"protocol", "baselineSearchDigest", "maximumAddedResources", "maximumQuantityPerBall", "maximumReceipts", "unitCostCeilings", "calibrationSetDigest", "calibrations", "signer", "digest"}, "Metabolic allowance")
    _check(allowance["protocol"] == "jevyr.metabolic-allowance/1" and allowance["baselineSearchDigest"] == seal["searchDigest"] and _digest(allowance["digest"]) and _hash({k:v for k,v in allowance.items() if k != "digest"}) == allowance["digest"], "Metabolic allowance digest or baseline differs")
    _check(_integer(allowance["maximumQuantityPerBall"], 1, 64) and _integer(allowance["maximumReceipts"], 1, 4096), "Invalid metabolic bounds")
    _vector(allowance["maximumAddedResources"])
    ceilings = _exact(allowance["unitCostCeilings"], set(KINDS), "Metabolic ceilings")
    for kind in KINDS:
        _vector(ceilings[kind])
        _check(all(ceilings[kind][key] <= allowance["maximumAddedResources"][key] for key in RESOURCES), "Dose exceeds allowance")
    calibrations = allowance["calibrations"]
    _check(isinstance(calibrations, list) and len(calibrations) <= 5, "Unbounded metabolic calibration set")
    seen: set[str] = set()
    for value in calibrations:
        item = _exact(value, {"kind", "digest", "summaryDigest", "maximumDose"}, "Calibration binding")
        _check(_digest(item["summaryDigest"]), "Invalid calibration summary digest")
        _check(item["kind"] in KINDS and item["kind"] not in seen and _digest(item["digest"]) and _integer(item["maximumDose"], 2, 64), "Invalid calibration binding")
        seen.add(item["kind"])
    reports = sorted([{"kind": c["kind"], "digest": c["digest"]} for c in calibrations], key=lambda c:c["kind"])
    _check(_hash({"protocol":"jevyr.metabolic-calibration-set/1", "reports":reports}) == allowance["calibrationSetDigest"], "Calibration set digest mismatch")
    signer = _exact(allowance["signer"], {"keyId", "publicKeyPem"}, "Metabolic signer")
    _check(_digest(signer["keyId"]) and isinstance(signer["publicKeyPem"], str) and len(signer["publicKeyPem"]) <= 1024, "Invalid metabolic signer")
    return allowance

def verify_redemption(value: Any, allowance: Mapping[str, Any], seal: Mapping[str, Any], previous: Mapping[str, Any] | None = None) -> dict[str, Any]:
    from .client import _base64, _canonical, _dsse_pae, _loads_unambiguous, _public_key, _valid_canonical_timestamp
    from cryptography.exceptions import InvalidSignature
    result = _exact(value, {"protocol", "receipt", "envelope"}, "Metabolic redemption")
    _check(result["protocol"] == "jevyr.metabolism-redemption/1", "Invalid metabolic redemption protocol")
    receipt = _exact(result["receipt"], {"protocol", "caseId", "caseDigest", "runDigest", "policyDigest", "searchDigest", "allowanceDigest", "calibrationDigest", "ballId", "kind", "quantity", "sequence", "previousReceiptDigest", "issuedAt", "grant", "cumulativeGrant", "cumulativeQuantities", "work", "baselineUnchanged", "verdictAuthority", "digest"}, "Metabolic receipt")
    _check(receipt["protocol"] == "jevyr.metabolism-receipt/1" and receipt["baselineUnchanged"] is True and receipt["verdictAuthority"] == "none", "Metabolic receipt changes authority")
    _check(all(receipt[key] == seal[key] for key in ("caseId", "caseDigest", "runDigest", "policyDigest", "searchDigest")) and receipt["allowanceDigest"] == allowance["digest"], "Metabolic receipt crosses the sealed identity")
    if previous:
        _check(all(previous[key] == receipt[key] for key in ("caseId", "caseDigest", "runDigest", "policyDigest", "searchDigest", "allowanceDigest")), "Metabolic predecessor crosses the sealed identity")
    assert_ball_id(receipt["ballId"]); assert_quantity(receipt["quantity"])
    kind = receipt["kind"]
    calibration = next((c for c in allowance["calibrations"] if c["kind"] == kind), None)
    _check(kind in KINDS and calibration is not None and receipt["calibrationDigest"] == calibration["digest"], "Receipt lacks its exact measured calibration")
    _check(_integer(receipt["sequence"], 1, allowance["maximumReceipts"]) and receipt["sequence"] == (previous["sequence"] if previous else 0) + 1 and receipt["previousReceiptDigest"] == (previous["digest"] if previous else None), "Metabolic receipt sequence has a gap or wrong predecessor")
    _check(receipt["quantity"] <= allowance["maximumQuantityPerBall"] and _valid_canonical_timestamp(receipt["issuedAt"]), "Invalid metabolic quantity or timestamp")
    _check(_hash({k:v for k,v in receipt.items() if k != "digest"}) == receipt["digest"], "Metabolic receipt digest mismatch")
    grant = _vector(receipt["grant"]); cumulative = _vector(receipt["cumulativeGrant"]); quantities = _vector(receipt["cumulativeQuantities"], KINDS); work = _vector(receipt["work"], WORK)
    for key in RESOURCES:
        _check(grant[key] == allowance["unitCostCeilings"][kind][key] * receipt["quantity"] and cumulative[key] == (previous["cumulativeGrant"][key] if previous else 0) + grant[key] and cumulative[key] <= allowance["maximumAddedResources"][key], "Metabolic resource addition exceeds sealed allowance")
    for name in KINDS:
        _check(quantities[name] == (previous["cumulativeQuantities"][name] if previous else 0) + (receipt["quantity"] if name == kind else 0), "Metabolic cumulative quantity differs")
    _check(quantities[kind] <= calibration["maximumDose"], "Metabolic quantity exceeds measured dose")
    _check(all(work[name] == (receipt["quantity"] if name == WORK[KINDS.index(kind)] else 0) for name in WORK), "Metabolic receipt changes finite work kind")
    envelope = _exact(result["envelope"], {"payloadType", "payload", "signatures"}, "Metabolic envelope")
    _check(envelope["payloadType"] == PAYLOAD_TYPE and isinstance(envelope["signatures"], list) and len(envelope["signatures"]) == 1, "Metabolic envelope requires its separate signer")
    signature = _exact(envelope["signatures"][0], {"keyid", "sig"}, "Metabolic signature")
    _check(signature["keyid"] == allowance["signer"]["keyId"], "Metabolic signature is not the sealed signer")
    payload = _base64(envelope["payload"], "Metabolic payload")
    _check(_canonical(_loads_unambiguous(payload)) == payload == _canonical(receipt), "Signed metabolic payload differs from receipt")
    key, _spki = _public_key({**allowance["signer"], "algorithm":"Ed25519"})
    _check("sha256:" + hashlib.sha256(_spki).hexdigest() == allowance["signer"]["keyId"], "Metabolic SPKI key identity mismatch")
    try:
        key.verify(_base64(signature["sig"], "Metabolic signature"), _dsse_pae(PAYLOAD_TYPE, payload))
    except InvalidSignature:
        _check(False, "Metabolic receipt signature failed")
    return result

def verify_offers(value: Any, allowance: Mapping[str, Any], seal: Mapping[str, Any]) -> dict[str, Any]:
    from .client import _utf16_length
    result = _exact(value, {"protocol", "caseId", "runDigest", "allowanceDigest", "baselineSearchDigest", "admission", "cumulativeGrant", "cumulativeQuantities", "offers", "receipts"}, "Metabolic offers")
    _check(result["protocol"] == "jevyr.metabolic-offers/1" and result["caseId"] == seal["caseId"] and result["runDigest"] == seal["runDigest"] and result["allowanceDigest"] == allowance["digest"] and result["baselineSearchDigest"] == seal["searchDigest"] and result["admission"] in ("open", "closed"), "Metabolic offers cross sealed identity")
    _check(isinstance(result["receipts"], list) and len(result["receipts"]) <= min(320, allowance["maximumReceipts"]), "Unbounded metabolic receipt history")
    previous = None
    for entry in result["receipts"]:
        previous = verify_redemption(entry, allowance, seal, previous)["receipt"]
    grant = _vector(result["cumulativeGrant"]); quantities = _vector(result["cumulativeQuantities"], KINDS)
    _check(all(grant[k] == (previous["cumulativeGrant"][k] if previous else 0) for k in RESOURCES) and all(quantities[k] == (previous["cumulativeQuantities"][k] if previous else 0) for k in KINDS), "Metabolic offers omit signed history")
    offers = result["offers"]
    _check(isinstance(offers, list) and len(offers) <= 5 and (result["admission"] == "open" or not offers), "Invalid metabolic admission")
    seen_kinds: set[str] = set(); seen_balls: set[str] = set()
    for raw in offers:
        offer = _exact(raw, {"ballId", "kind", "minQuantity", "maxQuantity", "promisedEffect", "unitCostCeiling", "unitWork", "calibration"}, "Metabolic offer")
        assert_ball_id(offer["ballId"]); assert_quantity(offer["maxQuantity"])
        kind = offer["kind"]
        _check(kind in KINDS and kind not in seen_kinds and offer["ballId"] not in seen_balls and offer["minQuantity"] == 1, "Duplicate or invalid metabolic offer")
        seen_kinds.add(kind); seen_balls.add(offer["ballId"])
        calibration = _exact(offer["calibration"], {"digest", "scope", "pairedSeeds", "doses", "monotonic", "identicalEvidenceJudgment", "recallDifference", "reproducibilityDifference", "nonInferiorityMargin", "maximumObservedUnitResources"}, "Metabolic calibration summary")
        binding = next((c for c in allowance["calibrations"] if c["kind"] == kind), None)
        _check(binding is not None and _hash(calibration) == binding["summaryDigest"], "Offered calibration summary differs from sealed digest")
        _check(binding is not None and calibration["digest"] == binding["digest"] and offer["maxQuantity"] <= min(allowance["maximumQuantityPerBall"], binding["maximumDose"] - quantities[kind]) and calibration["monotonic"] is True and calibration["identicalEvidenceJudgment"] is True and _integer(calibration["pairedSeeds"], 1), "Offer has no exact measured calibration")
        _check(isinstance(calibration["scope"], str) and 8 <= _utf16_length(calibration["scope"]) <= 2048 and isinstance(calibration["doses"], list) and len(calibration["doses"]) <= 65 and all(_integer(d, 0, 64) for d in calibration["doses"]), "Invalid calibration sample")
        for field in ("recallDifference", "reproducibilityDifference"):
            interval = _exact(calibration[field], {"lower", "upper", "confidence"}, "Calibration interval")
            _check(type(interval["lower"]) in (int, float) and type(interval["upper"]) in (int, float) and -1 <= interval["lower"] <= interval["upper"] <= 1 and interval["confidence"] == 0.95, "Invalid calibration interval")
        _check(type(calibration["nonInferiorityMargin"]) in (int, float) and 0 <= calibration["nonInferiorityMargin"] <= 0.1 and isinstance(offer["promisedEffect"], str) and 0 < len(offer["promisedEffect"]) <= 1024, "Invalid finite effect promise")
        costs = _vector(offer["unitCostCeiling"]); work = _vector(offer["unitWork"], WORK)
        observed_costs = _vector(calibration["maximumObservedUnitResources"])
        _check(all(observed_costs[k] <= costs[k] for k in RESOURCES), "Offer understates measured unit resources")
        _check(all(costs[k] == allowance["unitCostCeilings"][kind][k] and costs[k] * offer["maxQuantity"] + grant[k] <= allowance["maximumAddedResources"][k] for k in RESOURCES), "Offer exceeds additive allowance")
        _check(all(work[k] == (1 if k == WORK[KINDS.index(kind)] else 0) for k in WORK), "Offer changes work kind")
    return result
